### DISCUSSION TITLE
Questions about JMH, vector API

### DISCUSSION THREAD

**Discussion by bobtins** [gunnarmorling/1brc#500] [General]
I've been mostly lurking, but I have lots of experience making Java go fast with big data, so I've been toying with various approaches.
I have a couple questions:

- I did a search for JMH in discussions but got no hits; is no one using it?
- I was looking at using Vector API but it seems like Graal isn't supporting it yet, so I wonder if that's a strike against it

My idea was to use Vector API to get the fattest SIMD registers possible, and try to get the JDK to generate some reasonable AVX2 instructions. My desktop (Ryzen 1700, 8 Zen 1 cores, 32G RAM) is coincidentally not too far from the target VM (8 Zen 2 cores and 32G RAM). It seems like JMH would be a great way of identifying hotspots so I'm surprised not to see it being used to optimize algorithms.

**Comment by LamGC** [comment]
In my opinion, JMH might not be the right tool for 1BRC. Given that I/O is a major factor, the high-precision nature of JMH tends to become inconsistent. Instead of looking for a micro-benchmark of the whole app, most developers would prefer using perf to see what’s actually happening under the hood (CPU cycles, cache misses, etc.) where the real optimizations happen.

> As a latecomer to this challenge, I hope these observations provide some helpful context for those who come across this later.
